sdks/python/apache_beam/runners/interactive/caching/datatype_inference.py

```diff
@@ -0,0 +1,109 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import absolute_import
+
+import array
+import json
+from collections import OrderedDict
+
+import numpy as np
+import pyarrow as pa
+from past.builtins import unicode
+
+from apache_beam.typehints import trivial_inference
+from apache_beam.typehints import typehints
+
+# pylint: disable=wrong-import-order, wrong-import-position
+try:
+  from avro.schema import Parse  # avro-python3 library for python3
+except ImportError:
+  from avro.schema import parse as Parse  # avro library for python2
+# pylint: enable=wrong-import-order, wrong-import-position
+
+
+def infer_element_type(elements):
+  element_type = typehints.Union[[
+      trivial_inference.instance_to_type(e) for e in elements
+  ]]
+  return element_type
+
+
+def infer_typehints_schema(data):
+  """
+  TODO: For internal use only.
+  """
+  column_data = OrderedDict()
+  for row in data:
+    for key, value in row.items():
+      column_data.setdefault(key, []).append(value)
+  column_types = OrderedDict([
+      (key, infer_element_type(values)) for key, values in column_data.items()
+  ])
+  return column_types
+
+
+def infer_avro_schema(data, use_fastavro=False):
+  """
+  TODO: For internal use only.
+  """
+  _typehint_to_avro_type = {
+      typehints.Union[[int]]: "int",
```

Comment:
`typehints.Union[[int]]` returns a plain `int`, so I believe it would be clearer to write `int: "int",`.
You could also code all the Union cases:
```py
def typehint_to_avro_type(value):
  if instance(value, typehints.UnionConstraint):
    return [typehint_to_avro_type(union_type) for union_type in value.union_types]
  else:
    return _typehint_to_avro_type[value]
```

Reply:
Thanks, the code is much cleaner with this helper function.